### ISSUE TITLE
Consider incorporating required parameter validation into the parameterized constructor delegate

### ISSUE THREAD

**Issue by eiriktsarpalis** [eiriktsarpalis/PolyType#78]
Today required parameters are [only surfaced as metadata](https://eiriktsarpalis.github.io/PolyType/api/PolyType.Abstractions.IConstructorParameterShape.html#PolyType_Abstractions_IConstructorParameterShape_IsRequired) in the PolyType abstractions without any validation taking place in the constructor delegate. The responsibility of validation is then pushed to library authors who need to separately keep track of what parameters have been set.

Given that the constructor argument states [already keep track of what parameters have been set](https://eiriktsarpalis.github.io/PolyType/api/PolyType.Abstractions.IConstructorParameterShape.html#PolyType_Abstractions_IConstructorParameterShape_IsRequired) it should be possible to bake in validation in the constructor delegate itself. This could either be the default behavior or perhaps exposed as a secondary helper predicate in the shape of `Func<TArgumentState, bool>` which should let users craft bespoke exceptions.

Another alternative is adding providing a poolable `BitArray` in the `TypeShape.Utilities` namespace that implements common required parameter validation workflows.

cc @AArnott

**Comment by AArnott** [comment]
Seems like #140 is a proposal to solve this issue. Should we close one of them?

**Comment by eiriktsarpalis** [comment]
Yes, this is clearly both a duplicate and less specific than what #140 proposes. Closing.
